### PR TITLE
[WIP] Suppress error messages for missing raster tiles

### LINKED ISSUE #160: Suppress error messages for missing raster tiles
Returning http status code 204 (no content) with no body for missing tiles is used in many mapserver implementations like TileServer GL (https://github.com/maptiler/tileserver-gl/issues/46) and also proposed by the OGC API Tiles (https://github.com/opengeospatial/ogcapi-tiles/blob/master/core/standard/requirements/core/REQ_tc-success.adoc).    
This is particularly the case for raster tile sources  where heterogenous maximum zoom levels are common.   
MapLibre doesn't explicitly handle the 204 status for raster tiles which leads to a large number of error messages in the dev console.   

### Expected Behavior
No errors are thrown when a mapserver returns 204 for a missing raster tile and therefore no error messages in the dev console

### Actual Behavior
image

### PR DESCRIPTION
**Pull request by mactrem**
This PR adds explicit support for the 204 status code returned from a tileserver to get rid of 
the large number of error messages in the dev console when requesting raster sources.
  
closes #160 

## Launch Checklist

 - [ ] confirm your changes do not include backports from Mapbox projects (unless with compliant license)
 - [ ] briefly describe the changes in this PR
 - [ ] include before/after visuals or gifs if this PR includes visual changes
 - [ ] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [ ] post benchmark scores
 - [ ] manually test the debug page
 - [ ] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [ ] add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`

### CODE REVIEW

File: src/util/ajax.js [hunk 1]
```diff
@@ -215,7 +215,7 @@ function makeXMLHttpRequest(requestParameters: RequestParameters, callback: Resp
         callback(new Error(xhr.statusText));
     };
     xhr.onload = () => {
-        if (((xhr.status >= 200 && xhr.status < 300) || xhr.status === 0) && xhr.response !== null) {
+        if (((xhr.status >= 200 && xhr.status < 300 && xhr.status != 204) || xhr.status === 0) && xhr.response !== null) {
```
Comment (wipfli) [inline src/util/ajax.js:218]: why not `!==`?

Comment (mactrem) [reply src/util/ajax.js]: sure, careless mistake

**Comment by lseelenbinder** [comment]
Thanks @mactrem! I like the change, but it appears to have broken some unit tests.

Would you be able to take a look at the test failures and fix the failing tests?

**Comment by mactrem** [comment]
> Thanks @mactrem! I like the change, but it appears to have broken some unit tests.
> 
> Would you be able to take a look at the test failures and fix the failing tests?

@lseelenbinder 
I have the problem that the unit tests currently are not running on a windows machine. So i try to fix that first and then continue with the broken test.

**Comment by HarelM** [comment]
@mactrem what's the status of this PR? any chance you can fix the tests? I would like to merge this, I don't see an issue with the code as long as the tests pass.
It might be interesting to add a test to cover this case somehow...
Also don't forget to merge from `main`.

**Comment by HarelM** [comment]
We've just merged the typescript branch to `main`, so this basically needs to be done on the new branch as the file changed.
Once all the tests are passing please change the changelog file with this addition.
Please let us know if this is possible for you to do, I would like to get this into version 2.x.

**Comment by xabbu42** [comment]
I rebased this branch on current main at https://github.com/xabbu42/maplibre-gl-js/tree/pull-160 and the `npm run test` passes. that said, I believe this fix is wrong as it treats 204 the same as 404. According to my reading of https://github.com/opengeospatial/ogcapi-tiles/blob/master/core/standard/requirements/core/REQ_tc-success.adoc it should be treated like an empty image with a 200 http result code. I also would expect 204 to be handled as an empty but existing vector tile for vector layers, same as 200 http result code with a content-length of 0.

**Comment by HarelM** [comment]
I'm not sure what are you suggesting @xabbu42, if you think this is not the right direction in terms of fixing this issue we can and should reject this PR.
I don't have strong feelings wither way...

**Comment by xabbu42** [comment]
After reading through https://github.com/mapbox/mapbox-gl-js/issues/1800 this seems to actually implement the approach favored there, that is use 204 instead of 404 explicitely to allow for missing tiles (and the according behaviour of maplibre-gl-js to fall back on other zoom levels etc) without triggering the browser loging of 404.

**Comment by xabbu42** [comment]
@HarelM I'm not suggesting anything yet as I'm still confused on the situation myself. My gut feeling would be that 204 should be equivalent to 200 with content-length 0 at least for vector tiles. I can see the logic behind treating 204 as an empty image for raster tiles.  But the discussion in the mapbox issue seems to disagree with this takes.

**Comment by xabbu42** [comment]
In the end compatiblity to other products fetching tiles is probably the most important concern here. So this pull request should only make it in if other consumers and especially mapbox-gl-js 2.x have the same behavior, that is they also fallback to higher zoomlevels on 204 responses.
